File: architecture/threads_attack.rb
# from http://blog.self-loop.com/architecture_lecture/html/09_locking.html
# from http://blog.self-loop.com/architecture_lecture/html/10_ruby_mutex.html

# Run in Jruby to simulate a data race
# If you have two threads working and changing the same data, you want
# to use locks to prevent the threads overwriting each other.

require "concurrent"

NUM_ITERS = 1_000_000

class MyMutex
  def initialize
    @thread1_wants_to_lock = Concurrent::AtomicBoolean.new(false)
    @thread2_wants_to_lock = Concurrent::AtomicBoolean.new(false)
    @turn = Concurrent::AtomicFixnum.new(0)
  end

  def lock(thread_idx )
    if thread_idx == 1
      @thread1_wants_to_lock.make_true
      @turn.value = 2
      while @thread2_wants_to_lock.true? && @turn.value == 2
        # Busy loop AKA "Spinlock"
      end
    else
      @thread2_wants_to_lock.make_true
      @turn.value = 1
      while @thread1_wants_to_lock.true? && @turn.value == 1
        # Busy loop AKA "Spinlock"
      end
    end
  end

  def unlock(thread_idx)
    if thread_idx == 1
      @thread1_wants_to_lock.make_false
    else
      @thread2_wants_to_lock.make_false
    end
  end

end

i = 0
lock_for_i = MyMutex.new

thr1 = Thread.new do
  NUM_ITERS.times do
    lock_for_i.lock(1)
    i += 1
    lock_for_i.unlock(1)
  end
end

thr2 = Thread.new do
  NUM_ITERS.times do
    lock_for_i.lock(2)
    i += 1
    lock_for_i.unlock(2)
  end
end

thr1.join
thr2.join

puts i

File: architecture/threads_attack.c
// from http://blog.self-loop.com/architecture_lecture/html/11_c_lock.html

#include <pthread.h>
#include <stdbool.h>
#include <stdio.h>

const int NUM_ITERS = 1000000;
const int NUM_THREADS = 2;

// Using the _Atomic keyword prevents data race conditions
typedef struct
{
  _Atomic bool thread0WantsLock;
  _Atomic bool thread1WantsLock;
  _Atomic int turn;
} MyLock;

void acquireLock(MyLock *lock, int threadIdx)
{
  if (threadIdx == 0)
  {
    lock->thread0WantsLock = true;
    lock->turn = 1;
    while (lock->thread1WantsLock && (lock->turn == 1))
    {
      // Spinlock
    }
  }
  else
  {
    lock->thread1WantsLock = true;
    lock->turn = 0;
    while (lock->thread0WantsLock && (lock->turn == 0))
    {
      // Spinlock
    }
  }
}

void releaseLock(MyLock *lock, int threadIdx)
{
  if (threadIdx == 0)
  {
    lock->thread0WantsLock = false;
  }
  else
  {
    lock->thread1WantsLock = false;
  }
};

typedef struct
{
  int threadIdx;
  int *result;
  MyLock *lock;
} ThreadArgument;

void *thread_work(void *_argument)
{
  // Cast.
  ThreadArgument *threadArgument = (ThreadArgument *)_argument;

  for (int i = 0; i < NUM_ITERS; i++)
  {
    acquireLock(threadArgument->lock, threadArgument->threadIdx);
    *(threadArgument->result) += 1;
    releaseLock(threadArgument->lock, threadArgument->threadIdx);
  }

  return NULL;
}

int main()
{
  pthread_t threads[NUM_THREADS];
  int result = 0;
  MyLock lock = {
      .thread0WantsLock = false,
      .thread1WantsLock = false,
      .turn = -1};

  ThreadArgument arg0 = {
      .threadIdx = 0,
      .result = &result,
      .lock = &lock};

  ThreadArgument arg1 = {
      .threadIdx = 1,
      .result = &result,
      .lock = &lock};

  pthread_create(&threads[0], NULL, thread_work, &arg0);
  pthread_create(&threads[1], NULL, thread_work, &arg1);

  pthread_join(threads[0], NULL);
  pthread_join(threads[1], NULL);

  printf("Result: %d\n", result);

  return 0;
}
